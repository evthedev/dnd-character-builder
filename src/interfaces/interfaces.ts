import { SELECT_TYPE } from "../common/constants"

export type ISelectType= SELECT_TYPE.Class | SELECT_TYPE.Subclass | null

export interface IGenericEntity {
	index: string;
	name: string
}

export interface IAntSelectEvent {
	value: string;
	label: string;
	key: string
}

export interface ICharacter extends IGenericEntity {
	desc: string[]
}

export interface ISpellsPanelProps {
	spells: IGenericEntity[];
	title: string;
}

export interface IState {
	selectType: ISelectType;
	characterClasses: ICharacter[];
	characterSubclasses: ICharacter[];
	selectedCharacter: any; // needs typing
	selectedClassSpells: IGenericEntity[];
	selectedSubclassSpells: IGenericEntity[];
	selectedSubclassLevel: IGenericEntity | null;
	selectedSubclassLevelSpells: IGenericEntity[];
}