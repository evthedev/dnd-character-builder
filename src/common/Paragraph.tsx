import styled from 'styled-components';

export const Paragraph = styled.p`
	font-size: 18px;
	color: #666;
`;