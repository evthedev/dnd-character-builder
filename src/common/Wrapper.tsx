import styled from 'styled-components';

export const Wrapper = styled.div`
	height: 100vh;
	width: 100vw;
	background-color: aliceblue;
	display: flex;
	flex-direction: column;
`;