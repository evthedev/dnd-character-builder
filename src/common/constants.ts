export enum SELECT_TYPE {
	Class = 'class',
	Subclass = 'subclass'
}

export const selectTypes = [
	SELECT_TYPE.Class,
	SELECT_TYPE.Subclass
]
export const API_ENDPOINT_BASE = 'https://www.dnd5eapi.co';
export const API_ENDPOINT_CLASSES = API_ENDPOINT_BASE + '/api/classes';
export const API_ENDPOINT_SUBCLASSES = API_ENDPOINT_BASE + '/api/subclasses';
