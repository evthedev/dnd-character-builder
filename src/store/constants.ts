import { ICharacter, ISelectType, IGenericEntity } from '../interfaces/interfaces';

export const SET_SELECT_TYPE = 'SET_SELECT_TYPE';
export const SET_CLASSES = 'SET_CLASSES';
export const SET_SUBCLASSES = 'SET_SUBCLASSES';
export const SET_SELECTED_CHARACTER = 'SET_SELECTED_CHARACTER';
export const GET_SELECTED_CHARACTER = 'GET_SELECTED_CHARACTER';
export const CLEAR_SELECTED_CHARACTER = 'CLEAR_SELECTED_CHARACTER';
export const SET_SELECTED_CLASS_SPELLS = 'SET_SELECTED_CLASS_SPELLS';
export const GET_SUBCLASSES = 'GET_SUBCLASSES';

export interface SetSelectTypeAction {
	type: typeof SET_SELECT_TYPE;
	payload: ISelectType
}

export interface SetClassesAction {
	type: typeof SET_CLASSES;
	payload: ICharacter[]
}

export interface SetSubclassesAction {
	type: typeof SET_SUBCLASSES;
	payload: ICharacter[]
}

export interface SetSelectedCharacterAction {
	type: typeof SET_SELECTED_CHARACTER;
	payload: ICharacter
}

export interface GetSelectedCharacterAction {
	type: typeof GET_SELECTED_CHARACTER;
	payload: ICharacter
}

export interface ClearSelectedCharacterAction {
	type: typeof CLEAR_SELECTED_CHARACTER;
}

export interface SetSelectedClassSpellsAction {
	type: typeof SET_SELECTED_CLASS_SPELLS;
	payload: IGenericEntity[]
}

export type actionTypes = SetSelectTypeAction | SetClassesAction | SetSubclassesAction | SetSelectedCharacterAction | GetSelectedCharacterAction | ClearSelectedCharacterAction | SetSelectedClassSpellsAction