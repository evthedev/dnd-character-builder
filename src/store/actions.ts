import { SET_CLASSES, SET_SELECTED_CHARACTER, SET_SELECT_TYPE, SET_SUBCLASSES, actionTypes, CLEAR_SELECTED_CHARACTER, SET_SELECTED_CLASS_SPELLS } from "./constants";
import { ISelectType, IGenericEntity, ICharacter, IAntSelectEvent } from '../interfaces/interfaces';
import { Dispatch } from "redux";
import { API_ENDPOINT_CLASSES, API_ENDPOINT_SUBCLASSES, API_ENDPOINT_BASE } from '../common/constants';

export const setSelectType = (selectType: ISelectType): actionTypes => ({
	type: SET_SELECT_TYPE,
	payload: selectType
})

export const setClasses = (classes: ICharacter[]): actionTypes => ({
	type: SET_CLASSES,
	payload: classes
})

export const setSubclasses = (subclasses: ICharacter[]): actionTypes => ({
	type: SET_SUBCLASSES,
	payload: subclasses
})

export const setSelectedCharacter = (selectedCharacter: ICharacter): actionTypes => ({
	type: SET_SELECTED_CHARACTER,
	payload: selectedCharacter
})

export const clearSelectedCharacter = (): actionTypes => ({
	type: CLEAR_SELECTED_CHARACTER,
})

export const setSelectedClassSpells = (results: IGenericEntity[]): actionTypes => ({
	type: SET_SELECTED_CLASS_SPELLS,
	payload: results
})

export const fetchClassesFromApi = () => (
	async (dispatch: Dispatch) => {
		const response = await fetch(API_ENDPOINT_CLASSES, {
			method: 'GET'
		});
		const classesResponse = await response.json();
		dispatch(setClasses(classesResponse.results));
	}
)

export const fetchSubclassesFromApi = () => (
	async (dispatch: Dispatch) => {
		const response = await fetch(API_ENDPOINT_SUBCLASSES, {
			method: 'GET'
		});
		const subclassesResponse = await response.json();
		dispatch(setSubclasses(subclassesResponse.results));
	}
)

export const loadSelectedCharacter = (selected: IAntSelectEvent) => (
	async (dispatch: Dispatch) => {
		const response = await fetch(API_ENDPOINT_BASE + selected.value, {
			method: 'GET'
		});
		const responseJson = await response.json();
		return dispatch(setSelectedCharacter(responseJson))
	}
)

export const loadClassSpells = (selectedCharacter: any) => (
	async (dispatch: Dispatch) => {
		if (selectedCharacter.spells && typeof selectedCharacter.spells === 'string') {
			const response = await fetch(API_ENDPOINT_BASE + selectedCharacter.spells, {
				method: 'GET'
			});
			const responseJson = await response.json();
			return dispatch(setSelectedClassSpells(responseJson.results))
		}
	}
)

export const clearClassSpells = () => (
	async (dispatch: Dispatch) => {
		return dispatch(setSelectedClassSpells([]))
	}
)
