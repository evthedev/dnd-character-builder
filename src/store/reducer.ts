import { defaultState } from '../common/defaults';
import { IState } from '../interfaces/interfaces';
import { SET_SELECT_TYPE, SET_CLASSES, SET_SUBCLASSES, SET_SELECTED_CHARACTER, CLEAR_SELECTED_CHARACTER, actionTypes, SET_SELECTED_CLASS_SPELLS } from './constants';

export const reducer = (state: IState = defaultState, action: actionTypes): IState => {
	switch (action.type) {
		case SET_SELECT_TYPE:
			return {
				...state,
				selectType: action.payload
			}
		case SET_CLASSES:
			return {
				...state,
				characterClasses: action.payload
			}
		case SET_SUBCLASSES:
			return {
				...state,
				characterSubclasses: action.payload
			}
		case SET_SELECTED_CHARACTER:
			return {
				...state,
				selectedCharacter: action.payload
			}
		case CLEAR_SELECTED_CHARACTER:
			return {
				...state,
				selectedCharacter: null
			}
		case SET_SELECTED_CLASS_SPELLS:
			return {
				...state,
				selectedClassSpells: action.payload
			}
		default:
			return state		
	}
}


// selectionType: ISelectType | null,
// 	characterClasses: string[],
// 	characterSubclasses: string[],
// 	selectedCharacterClassOrSubclass: IClassOrSubclass | {},
// 	selectedClassSpells: IGenericEntity[],
// 	selectedSubclassSpells: IGenericEntity[],
// 	selectedSubclassLevel: IGenericEntity | {},
// 	selectedSubclassLevelSpells