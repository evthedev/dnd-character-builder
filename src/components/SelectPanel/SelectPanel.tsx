import React, { Fragment } from 'react';

import { Select, Card, Space } from 'antd';
import { selectTypes, SELECT_TYPE } from '../../common/constants';
import { useDispatch, useSelector } from 'react-redux';
import { clearSelectedCharacter, fetchClassesFromApi, fetchSubclassesFromApi, loadSelectedCharacter, setSelectType } from '../../store/actions';
import { IAntSelectEvent, IState } from '../../interfaces/interfaces';

const { Option } = Select;

export const SelectPanel: React.FC = React.memo((): JSX.Element => {
	const dispatch = useDispatch();

	const selectType = useSelector((state: IState) => state.selectType)
	const classes = useSelector((state: IState) => state.characterClasses)
	const subclasses = useSelector((state: IState) => state.characterSubclasses)

	const handleChange = (selectedType: any): void => {
		const selectedTypeValue = selectedType.value;
		dispatch(setSelectType(selectedTypeValue));
		dispatch(clearSelectedCharacter());
		if (selectedTypeValue === SELECT_TYPE.Class && classes.length === 0) {
			dispatch(fetchClassesFromApi());
		} else if (selectedTypeValue === SELECT_TYPE.Subclass && subclasses.length === 0) {
			dispatch(fetchSubclassesFromApi());
		}
	}

	const handleCharacterSelect = (selected: IAntSelectEvent) => {
		dispatch(loadSelectedCharacter(selected)); 
	}

	return (
		<Fragment>
			<Space direction="vertical" style={{ width: '100%' }}>
				<Card size="small" title="Select Character Type" style={{ width: '100%' }}>
					<Select style={{ width: '100%' }} labelInValue={true} onChange={handleChange}>
						{ selectTypes.map((item: any) => (
							<Option value={item} key={item}>{ item }</Option>
						)) }
					</Select>
				</Card>
			</Space>			
			{ selectType &&
				<Space direction="vertical" style={{ width: '100%' }}>
					<Card size="small" title={'Select ' + selectType} style={{ width: '100%' }}>
						<Select style={{ width: '100%' }} labelInValue={true} onChange={handleCharacterSelect}>
						{ selectType as SELECT_TYPE === SELECT_TYPE.Class &&
							(classes).map((item: any) => (
								<Option value={item.url} key={item.index}>{ item.name }</Option>
							))
						}
						{ selectType as SELECT_TYPE === SELECT_TYPE.Subclass &&
							(subclasses).map((item: any) => (
								<Option value={item.url} key={item.index}>{ item.name }</Option>
							))
						}
						</Select>
					</Card>
				</Space>
			}
		</Fragment>
	)
})