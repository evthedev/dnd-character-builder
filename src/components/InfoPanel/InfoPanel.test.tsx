import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import { Wrapper } from '../../common/Wrapper';
import { Paragraph } from '../../common/Paragraph';
import { createStore } from 'redux';
import { reducer } from '../../store/reducer';
import { Provider } from 'react-redux';
import { IState } from '../../interfaces/interfaces';
import { SelectPanel } from './SelectPanel';
import { Select } from 'antd';
import { setSelectType } from '../../store/actions';
import { SELECT_TYPE } from '../../common/constants';

Enzyme.configure({ adapter: new EnzymeAdapter()});

describe.skip('<SelectPanel /> component', () => {
	let wrapper: any = null;
	let mockStoreForDispatch: any = '';

    const mockInitialState: IState = {
        selectType: null,
        characterClasses: [],
        characterSubclasses: [],
        selectedCharacter: null,
        selectedClassSpells: [],
        selectedSubclassSpells: [],
        selectedSubclassLevel: null,
		selectedSubclassLevelSpells: [],
	}

	const mockInitialStateWithClasses: IState = {
		...mockInitialState,
		selectType: SELECT_TYPE.Class,
		characterClasses: [
			{
				index: 'class-1',
				name: 'Class 1',
				desc: ['Description of Class 1']
			},
			{
				index: 'class-2',
				name: 'Class 2',
				desc: ['Description of Class 2']
			},
		]
	}
	
    const getWrapper = (mockStore = createStore(reducer, mockInitialState)) => mount(
        <Provider store={mockStore}>
            <SelectPanel />
        </Provider>
    )
    
    // beforeEach(() => {
	// 	mockStoreForDispatch = createStore(reducer, mockInitialState);
	// 	mockStoreForDispatch.dispatch = jest.fn();
    //     wrapper = getWrapper(mockStoreForDispatch);
    // });

	it('should mount correctly', () => {

		const mockStore = createStore(reducer, mockInitialState);
        wrapper = getWrapper(mockStore);

        expect(wrapper.isEmptyRender()).toBe(false);
		expect(wrapper).toMatchSnapshot();
	})

	it('should render only one <Select> component', () => {
		const mockStore = createStore(reducer, mockInitialState);
		// mockStore.dispatch = jest.fn();
        wrapper = getWrapper(mockStore);

		expect(wrapper.find(Select).length).toBe(1);
		
		
		// wrapper.find(Select).simulate('click');
		// expect(wrapper.find('.ant-select-open').length).toBe(1);
		// expect(wrapper.find('.ant-select-dropdown').not.to.hasClass('ant-select-dropdown-hidden')).toBe(true);
		// expect(wrapper.find('.ant-select-dropdown').not.hasClass('ant-select-dropdown-hidden')).toBe(false);

		// wrapper.find('.rc-virtual-list-holder-inner .ant-select-item-option:nth-child(2)').simulate('click');

        // expect(mockStoreForDispatch.dispatch).toHaveBeenCalledWith(setSelectType(SELECT_TYPE.Subclass))
        
		
		// expect(wrapper.find('#select-class-subclass').length).toBe(0);
		

        // expect(wrapper.contains(Wrapper)).toEqual(true);
        // expect(wrapper.find(Paragraph).text()).toEqual('Filter by status');
        // expect(wrapper.contains(StatusFilter)).toEqual(true);
	})

	it('should render both <Select> components', () => {
		const mockStoreWithClasses = createStore(reducer, mockInitialStateWithClasses);
        wrapper = getWrapper(mockStoreWithClasses);
		expect(wrapper.find(Select).length).toBe(2);
	})
})