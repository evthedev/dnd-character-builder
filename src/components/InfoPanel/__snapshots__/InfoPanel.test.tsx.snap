// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectPanel /> component should mount correctly 1`] = `ReactWrapper {}`;
